### PR TITLE
8278602: CDS dynamic dump may access unloaded classes

### PR DESCRIPTION
**Pull request by iklam**
Cause of crash:

When dumping a CDS archive,  while iterating over entries of the `SystemDictionaryShared::_dumptime_table`, we do not check whether the classes are already unloaded. In the crash, we are trying to call `InstanceKlass::signer()` but the class has already been unloaded.

Fix:

Override the  template function `DumpTimeSharedClassTable::iterate` to ensure iteration safety.  Do not iterate over a class if its `class_loader_data` is no longer alive.

The assert in `DumpTimeSharedClassTable::IterationHelper` found another existing bug -- we were calling `SystemDictionaryShared::is_dumptime_table_empty()` without holding the `DumpTimeTable_lock`. I delayed the call until we have grabbed the lock.

Testing:

I have attached a test case into the bug report. Without the fix, it would reproduce the same crash in less than a minute. With the fix, the crash is no longer reproducible.

Unfortunately, the test case requires a ZGC patch (thanks to @stefank) that adds delays to  increase the likelihood of seeing unloaded classes inside the `_dumptime_table`. Therefore, I cannot integrate the test as a jtreg test. I'll mark the bug as **noreg-hard**

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278602](https://bugs.openjdk.java.net/browse/JDK-8278602): CDS dynamic dump may access unloaded classes

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to 8584389e9e0b40c3e0288ffec7ac9159143976d4
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to ea1f318b61f393cfb4ba2fd3b7bb10e1195fd1af

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6859/head:pull/6859` \
`$ git checkout pull/6859`

Update a local copy of the PR: \
`$ git checkout pull/6859` \
`$ git pull https://git.openjdk.java.net/jdk pull/6859/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6859`

View PR using the GUI difftool: \
`$ git pr show -t 6859`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6859.diff">https://git.openjdk.java.net/jdk/pull/6859.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/systemDictionaryShared.cpp [hunk 1]
```diff
@@ -187,6 +187,11 @@ void SystemDictionaryShared::start_dumping() {
   _dump_in_progress = true;
 }
 
+void SystemDictionaryShared::stop_dumping() {
+  assert_lock_strong(DumpTimeTable_lock);
+  _dump_in_progress = true;
+}
+
```
Comment (stefank) [inline src/hotspot/share/classfile/systemDictionaryShared.cpp:194]: Did you really intend to set _dump_in_progress to true in stop_dumping()? start_dumping() also sets it to true.

**Review by stefank — CHANGES_REQUESTED** [review]
I've reviewed the interaction of the klasses in the _dumptime_table with the new is_loader_alive() check. I don't know the reset of the CDS code to know if the other changes are correct or not. I spotted something that looks weird:

File: src/hotspot/share/cds/dumpTimeClassInfo.inline.hpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  ITER* _iter;
+public:
+  IterationHelper(ITER* iter) {
+    _iter = iter;
+  }
+  bool do_entry(InstanceKlass* k, DumpTimeClassInfo& info) {
+    assert(SafepointSynchronize::is_at_safepoint(), "invariant");
+    assert_lock_strong(DumpTimeTable_lock);
+    if (k->is_loader_alive()) {
+      assert(k->is_loader_alive(), "must be");
```
Comment (coleenp) [inline src/hotspot/share/cds/dumpTimeClassInfo.inline.hpp:53]: This does seem a bit paranoid and redundant here.

File: src/hotspot/share/cds/dumpTimeClassInfo.inline.hpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      bool result = _iter->do_entry(k, info);
+      assert(k->is_loader_alive(), "must not change");
+      return result;
+    } else {
+      if (!SystemDictionaryShared::is_excluded_class(k)) {
```
Comment (coleenp) [inline src/hotspot/share/cds/dumpTimeClassInfo.inline.hpp:58]: I thought this was the original bug?  is_excluded_class() looks at mirror->signers() which if the class isn't alive, mirror->signers() will crash.  This has to be in the k->is_loader_alive() too.

**Review by coleenp — CHANGES_REQUESTED** [review]
Could you also add your unloads a lot test even though it doesn't reproduce this particular error without the ZGC change?  It might find a similar bug under stress conditions.

File: src/hotspot/share/cds/dumpTimeClassInfo.inline.hpp (hunk 3 above)
Comment (iklam) [reply src/hotspot/share/cds/dumpTimeClassInfo.inline.hpp:58]: is_excluded_class() only checks the DumpTimeClassInfo::_is_excluded field. It doesn't examine the mirror->signers(). The crash happened with SystemDictionaryShared::check_excluded_classes(), which does examine the signers.

```
bool SystemDictionaryShared::is_excluded_class(InstanceKlass* k) {
  assert(_no_class_loading_should_happen, "sanity");
  assert_lock_strong(DumpTimeTable_lock);
  Arguments::assert_is_dumping_archive();
  DumpTimeClassInfo* p = find_or_allocate_info_for_locked(k);
  return (p == NULL) ? true : p->is_excluded();
}
```

File: src/hotspot/share/cds/dumpTimeClassInfo.inline.hpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/cds/dumpTimeClassInfo.inline.hpp:53]: Oops, that's was left over code. I'll remove it.

File: src/hotspot/share/classfile/systemDictionaryShared.cpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp:194]: Oh, it should be `_dump_in_progress = false;`. I'll fix that.

**Comment by iklam** [comment]
> Could you also add your unloads a lot test even though it doesn't reproduce this particular error without the ZGC change? It might find a similar bug under stress conditions.

OK, I'll add the test case.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/test-classes/LotsUnloadApp.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    }
+}
+
+class DefinedAsHiddenKlass {
+    // ZGC region size is always a multiple of 2MB on x64.
+    // Make this slightly smaller than that.
+    static byte[] array = new byte[2 * 1024 * 1024 - 8 * 1024];
+    static String x;
+    static double d = 123;
+    static float f = 456;
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/test-classes/LotsUnloadApp.java:81]: Are the above declarations needed? They are not being used.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/test-classes/LotsUnloadApp.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    public DefinedAsHiddenKlass() {
+        // This will generate some lambda forms hidden classes for string concat.
+        x = "array size is "  + array.length + " bytes ";
+    }
+    public void doit(Runnable r) {
+        r.run();
+    }
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/test-classes/LotsUnloadApp.java:88]: I don't see the above method being called.

**Review by calvinccheung — COMMENTED** [review]
Looks good. Just couple of questions on the test.

File: src/hotspot/share/cds/dumpTimeClassInfo.inline.hpp (hunk 3 above)
Comment (coleenp) [reply src/hotspot/share/cds/dumpTimeClassInfo.inline.hpp:57]: Ok, sorry I got the names confused.

**Review by coleenp — APPROVED** [review]
Looks good. Thank you for adding the test.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/test-classes/LotsUnloadApp.java (hunk 5 above)
Comment (iklam) [reply test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/test-classes/LotsUnloadApp.java]: I've removed it.

File: test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/test-classes/LotsUnloadApp.java (hunk 4 above)
Comment (iklam) [reply test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/test-classes/LotsUnloadApp.java]: These are left overs. I've remove them.

**Review by calvinccheung — APPROVED** [review]

**Comment by iklam** [comment]
Thanks @stefank @calvinccheung @coleenp for the review. Latest version passed Mach5 CI tiers 1-4.
/integrate

**Pull request closed without merging**